ldms/src/sampler/zfs_zpool/zfs_zpool.c

```diff
@@ -0,0 +1,529 @@
+#define MIN_LAT_INDEX   10	/* minimum latency index 10 = 1024ns */
+#define POOL_IO_SIZE_MEASUREMENT        "zpool_io_size"
+#define MIN_SIZE_INDEX  9	/* minimum size index 9 = 512 bytes */
+
+typedef int (*stat_printer_f) (nvlist_t *, const char *, const char *);
+
+int complained_about_sync = 0;
+
+static ldmsd_msg_log_f log_fn;
+static base_data_t sampler_base;
+static libzfs_handle_t *g_zfs;
+
+static struct {
+	int vdev_list_idx;
+	int vdev_rec_idx;
+} index_store;
+
+static ldms_mval_t list_handle;
+
+typedef enum op_func_type {
+	FUNC_NOFUNCREQ = 0,
+	FUNC_SCRUB = 1,
+	FUNC_RESILVER = 2,
+	FUNC_REBUILD = 3,
+	FUNC_SCAN = 4
+} op_func_type_;
+
+static const char *const operation_types[] = {
+	[FUNC_NOFUNCREQ] = "none",
+	[FUNC_SCRUB] = "scrub",
+	[FUNC_RESILVER] = "resilver",
+	[FUNC_REBUILD] = "rebuild",
+	[FUNC_SCAN] = "scan"
+};
+
+/* metric templates for a zpool and scan status if any */
+static struct ldms_metric_template_s zfs_zpool[] = {
+	{"pool", 0, LDMS_V_CHAR_ARRAY, "", MAX_ZFS_NAME_LEN},
+	{"state", 0, LDMS_V_CHAR_ARRAY, "", MAX_ZFS_NAME_LEN},
+	{"total", 0, LDMS_V_U64, "", 1},
+	{"allocated", 0, LDMS_V_U64, "", 1},
+	{"free", 0, LDMS_V_U64, "", 1},
+	{"used", 0, LDMS_V_U64, "", 1},
+	{"scan_func", 0, LDMS_V_CHAR_ARRAY, "", MAX_ZFS_NAME_LEN},
+	{"scan_status", 0, LDMS_V_CHAR_ARRAY, "", MAX_ZFS_NAME_LEN},
+	{"scan_repaired", 0, LDMS_V_U64, "", 1},
+	{"scan_completed_in", 0, LDMS_V_U64, "", 1},
+	{"scan_errors", 0, LDMS_V_U32, "", 1},
+	{"scan_completed_on", 0, LDMS_V_U64, "", 1},
+	{0}
+};
+
+#define ZPOOL_METRICS_LEN (ARRAY_LEN(zfs_zpool) - 1)
+static int zpool_metric_ids[ZPOOL_METRICS_LEN];
+static size_t zpool_heap_sz;
+
+static int zpool_list_len = 0;	/* Aggregated number of vdev per zpool */
+
+/*****************************************************************************
+ * Initialize the structure as schema and add them to the base schema.
+ * Also calculate the size of memory needed per schema and add it to the ldms
+ * schema list.
+ ****************************************************************************/
+
+static int initialize_ldms_structs()
+{
+	/*ldms_record_t zpool_def;  a pointer */
+	ldms_record_t zpool_vdev_def;	/* a pointer */
+	int rc;
+
+	log_fn(LDMSD_LDEBUG, SAMP " initialize()\n");
+
+	/* Create the schema */
+	base_schema_new(sampler_base);
+	if (sampler_base->schema == NULL)
+		goto err1;
+
+	/* create the vdev record */
+	zpool_vdev_def = ldms_record_from_template("zfs_zpool_stats",
+						   zfs_zpool, zpool_metric_ids);
+	if (zpool_vdev_def == NULL)
+		goto err2;
+
+	zpool_heap_sz = ldms_record_heap_size_get(zpool_vdev_def);
+	rc = ldms_schema_record_add(sampler_base->schema, zpool_vdev_def);
+	if (rc < 0)
+		goto err3;
+
+	index_store.vdev_rec_idx = rc;
+	rc = zpool_iter(g_zfs, get_zpool_count, NULL);
+	/* add error for iter in case here */
+	rc = ldms_schema_metric_list_add(sampler_base->schema,
+					 "zpool_list",
+					 NULL, zpool_list_len * zpool_heap_sz);
+	if (rc < 0)
+		goto err2;
```

Comment:
`err3`?

Reply:
nope, if you read the comment, we delete the record manually only if it wasn't added to the metric. In that case it was done already.